libbpf-tools/tcprtt.bpf.c

```diff
@@ -9,49 +9,120 @@
 #include "bits.bpf.h"
 #include "maps.bpf.h"
 
+/* Taken from kernel include/linux/socket.h. */
+#define AF_INET		2	/* IP version 4 */
+#define AF_INET6	10	/* IP version 6 */
+
+
 const volatile bool targ_laddr_hist = false;
 const volatile bool targ_raddr_hist = false;
 const volatile bool targ_show_ext = false;
 const volatile __u16 targ_sport = 0;
 const volatile __u16 targ_dport = 0;
 const volatile __u32 targ_saddr = 0;
 const volatile __u32 targ_daddr = 0;
+const volatile __u8 targ_saddr_v6[IPV6_LEN] = {};
+const volatile __u8 targ_daddr_v6[IPV6_LEN] = {};
 const volatile bool targ_ms = false;
 
 #define MAX_ENTRIES	10240
 
 struct {
 	__uint(type, BPF_MAP_TYPE_HASH);
 	__uint(max_entries, MAX_ENTRIES);
-	__type(key, u64);
+	__type(key, struct hist_key);
 	__type(value, struct hist);
 } hists SEC(".maps");
 
 static struct hist zero;
 
+/*
+ * We cannot use the following:
+ * __builtin_memcmp(targ_*addr_v6, *, sizeof(targ_*addr_v6));
+ * Indeed, by using the builtin, we would discard the volatile qualifier of
+ * targ_*addr_v6, so the compiler would optimize it and replaces the call
```

Comment:
I don't understand the optimization mentioned here.

Reply:
@chenhengqi This is related to quirks in llvm compiler which I have not investigated it yet and I don't know whether upstream has any intention to change or not. Basically you have something like
```
const volatile __u8 targ_saddr_v6[IPV6_LEN] = {};
```

If we directly use targ_saddr_v6 in bpf program as most use cases in kernel bpf selftests, targ_saddr_v6 will not be treated as a 'const' and that is why its value can be changed after map is created but before program is loaded with libbpf skeleton.

But there is an abnormal here. If 'targ_saddr_v6' is used as a parameter to `__builtin_memcpy`, then the content of `target_saddr_v6` is treated as 'const' so the whole __builtin_memcpy(targ_saddr_v6, zero_addr_v6) is optimized away. 

I am aware of this issue and have not got chance to clarify this in upstream. I will check gcc behavior for this case as well and then will ask upstream.

Reply:
I filed an issue in llvm-project issue site: https://github.com/llvm/llvm-project/issues/63792
Feel to keep track of it or add a comment to it if you have any.